url748e34f967_Diab1o_java_7z_archiver_tgz-pJ8-com_swemel_sevenzip_TestJ8/src/com/swemel/sevenzip/compression/lz/BinTree.java

```diff
@@ -0,0 +1,372 @@
+                    distances[offset++] = _pos - curMatch2 - 1;
+                }
+            }
+            if (curMatch3 > matchMinPos) {
+                if (_bufferBase[_bufferOffset + curMatch3] == _bufferBase[cur]) {
+                    if (curMatch3 == curMatch2) {
+                        offset -= 2;
+                    }
+                    distances[offset++] = maxLen = 3;
+                    distances[offset++] = _pos - curMatch3 - 1;
+                    curMatch2 = curMatch3;
+                }
+            }
+            if (offset != 0 && curMatch2 == curMatch) {
+                offset -= 2;
+                maxLen = kStartMaxLen;
+            }
+        }
+
+        _hash[kFixHashSize + hashValue] = _pos;
+
+        int ptr0 = (_cyclicBufferPos << 1) + 1;
+        int ptr1 = (_cyclicBufferPos << 1);
+
+        int len0, len1;
+        len0 = len1 = kNumHashDirectBytes;
+
+        if (kNumHashDirectBytes != 0) {
+            if (curMatch > matchMinPos) {
+                if (_bufferBase[_bufferOffset + curMatch + kNumHashDirectBytes] !=
+                        _bufferBase[cur + kNumHashDirectBytes]) {
+                    distances[offset++] = maxLen = kNumHashDirectBytes;
+                    distances[offset++] = _pos - curMatch - 1;
+                }
+            }
+        }
+
+        int count = _cutValue;
+
+        while (true) {
+            if (curMatch <= matchMinPos || count-- == 0) {
+                _son[ptr0] = _son[ptr1] = kEmptyHashValue;
+                break;
+            }
+            int delta = _pos - curMatch;
+            int cyclicPos = ((delta <= _cyclicBufferPos) ?
+                    (_cyclicBufferPos - delta) :
+                    (_cyclicBufferPos - delta + _cyclicBufferSize)) << 1;
+
+            int pby1 = _bufferOffset + curMatch;
+            int len = Math.min(len0, len1);
+            if (_bufferBase[pby1 + len] == _bufferBase[cur + len]) {
+                while (++len != lenLimit) {
+                    if (_bufferBase[pby1 + len] != _bufferBase[cur + len]) {
+                        break;
+                    }
+                }
+                if (maxLen < len) {
+                    distances[offset++] = maxLen = len;
+                    distances[offset++] = delta - 1;
+                    if (len == lenLimit) {
+                        _son[ptr1] = _son[cyclicPos];
+                        _son[ptr0] = _son[cyclicPos + 1];
+                        break;
+                    }
+                }
+            }
+            if ((_bufferBase[pby1 + len] & 0xFF) < (_bufferBase[cur + len] & 0xFF)) {
+                _son[ptr1] = curMatch;
+                ptr1 = cyclicPos + 1;
+                curMatch = _son[ptr1];
+                len1 = len;
+            } else {
+                _son[ptr0] = curMatch;
+                ptr0 = cyclicPos;
+                curMatch = _son[ptr0];
+                len0 = len;
+            }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 240:250, 321:331.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.